docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,1385 @@
+- All methods MUST be `public`.
+- The method MUST use the constants defined in [dependency provider](#dependency-provider) when reaching out to dependencies.
+
+<details><summary markdown='span'>Additional Conventions for Project Development</summary>
+- Methods can have any access modifier.
+</details>
+
+## Gateway Controller
+
+**Description**
+
+`Gateway Controller` is special, reserved controller (see [Controllers](#controller)) in the framework, that serves as an entry point in `Zed` for serving `Client` requests (see [Zed Stub](#zed-stub) for more details).
+
+**Conventions**
+- `Gateway Controller` actions MUST define a single [transfer object](#transfer-object) as argument, and another/same [transfer object](#transfer-object) for return.
+
+## Mapper / Expander / Hydrator
+
+**Description**
+
+To differentiate between the recurring cases of data mapping, and unify naming, this convention aligns the terminology with intensions.
+
+`Persistance mapper` stands for a `mapper` that transforms [propel entity](#entity) or [entity transfers object](#transfer-object) into generic [transfer objects](#transfer-object).
+
+**Conventions**
+- Transforming / appending the provided data into another structure, is considered a `mapper` and the method MUST be named as `map[<SourceEntityName>[To<TargetEntityName>]]($sourceEntity, $targetEntity)`.
+  - Mappers MUST be free of any business logic, only mapping logic is allowed.
+  - Mapping MAY have multiple source entities if it still does NOT violate mapping-only directive.
+  - Mappers MUST NOT have data resolving (eg: remote calls, database lazy load) logic as they are utilised in high-batch processing scenarios.
+  - Mappers MUST NOT use dependencies or restrict to [module configuration](#module-configuration), other `mappers`, or lightweight [Services](#service-facade).
+  - `Persistence mappers` MUST be in [Persistence layer](#zed) only and named as `[Module|Entity]Mapper.php`.
+- Retrieving and appending additional data into the existing data structure, is considered an `expander` and the method MUST be named as `expand[With<AdditionalDataExplainer>]($targetEntity)`.
+- The `hydrator/hydration` keywords MUST NOT be used but instead `mapper` or `expander`.
+
+**Guidelines**
+- It is recommended to NOT define interface for `Persistence mappers`.
+
+## Model
+
+**Description**
+
+Models hold logic. Models MAY appear on multiple layers, the content must fit the corresponding [layer's responsibility](#layers).
+- [Communication layer](#application-layers) models are present in [Yves](#yves), [Glue](#glue), [Client](#client).
+- [Business layer](#application-layers) models are present in [Zed](#zed), [Service](#service).
+- [Persistence layer](#application-layers) models are present in [Zed](#zed).
+
+**Conventions**
+
+- Each model class MUST define and implement an interface (`[ModelName]Interface.php`) that holds the specification of each `public` method.
```

Comment:
Can't this be omitted in projects?

Reply:
yes, i've missed this. this is 1 of the very important project differences.